packages/mongodb-chatbot-server/src/routes/generateResponse.ts

```diff
@@ -0,0 +1,342 @@
+            choice.delta?.functionCall.arguments ?? ""
+          );
+        }
+      } else if (choice.message) {
+        logRequest({
+          reqId,
+          message: `Unexpected message in stream: no delta. Message: ${JSON.stringify(
+            choice.message
+          )}`,
+          type: "warn",
+        });
+      }
+    }
+    const shouldCallTool = functionCallContent.name !== "";
+    if (shouldCallTool) {
+      initialAssistantMessage.functionCall = functionCallContent;
+    }
+    newMessages.push(initialAssistantMessage);
+
+    logRequest({
+      reqId,
+      message: `LLM response: ${JSON.stringify(initialAssistantMessage)}`,
+    });
+    // Tool call
+    if (shouldCallTool) {
+      assert(
+        llm.callTool,
+        "You must implement the callTool() method on your ChatLlm to access this code."
+      );
+      const {
+        toolCallMessage,
+        references: toolReferences,
+        rejectUserQuery,
+      } = await llm.callTool({
+        messages: [...llmConversation, ...newMessages],
+        conversation,
+        dataStreamer,
+        request,
+      });
+      newMessages.push(convertMessageFromLlmToDb(toolCallMessage));
+
+      if (rejectUserQuery) {
+        newMessages.push({
+          role: "assistant",
+          content: noRelevantContentMessage,
+        });
+        dataStreamer.streamData({
+          type: "delta",
+          data: noRelevantContentMessage,
+        });
+      } else {
+        if (toolReferences) {
+          outputReferences.push(...toolReferences);
+        }
+        const answerStream = await llm.answerQuestionStream({
+          messages: [...llmConversation, ...newMessages],
+        });
+        const answerContent = await dataStreamer.stream({
+          stream: answerStream,
+        });
+        const answerMessage = {
+          role: "assistant",
+          content: answerContent,
+        } satisfies AssistantMessage;
+        newMessages.push(answerMessage);
+      }
+    }
+  } catch (err) {
+    const errorMessage =
+      err instanceof Error ? err.message : JSON.stringify(err);
+    logRequest({
+      reqId,
+      message: `LLM error: ${errorMessage}`,
+      type: "error",
+    });
+    logRequest({
+      reqId,
+      message: "Only sending vector search results to user",
+    });
+    const llmNotWorkingResponse = {
+      role: "assistant",
+      content: llmNotWorkingMessage,
+    } satisfies AssistantMessage;
+    dataStreamer.streamData({
+      type: "delta",
+      data: llmNotWorkingMessage,
+    });
+    newMessages.push(llmNotWorkingResponse);
+  }
+
+  // Stream back references
+  dataStreamer.streamData({
+    type: "references",
+    data: outputReferences,
+  });
+
+  // Add references to the last assistant message
+  (newMessages[newMessages.length - 1] as AssistantMessage).references =
+    outputReferences;
+
+  return { messages: newMessages };
+}
+export function convertMessageFromLlmToDb(
```

Comment:
Please always put 1 space between declarations/definitions
```suggestion

export function convertMessageFromLlmToDb(
```